### DISCUSSION TITLE
Table Inheritance: what is a leaf entity?

### DISCUSSION THREAD

**Discussion by stephanvierkant** [doctrine/doctrine2#8854] [Support Questions] [answered]
Interesting question I've found on StackOverflow: [What is a leaf entity in the inheritance hierarchy and what are examples of?](https://stackoverflow.com/q/68498891/1405981)

From [Doctrine's documentation](https://www.doctrine-project.org/projects/doctrine-orm/en/2.9/reference/inheritance-mapping.html#performance-impact):

> There is a general performance consideration with Class Table Inheritance: If the target-entity of a many-to-one or one-to-one association is a CTI entity, it is preferable for performance reasons that it be a leaf entity in the inheritance hierarchy, (ie. have no subclasses).
> Otherwise Doctrine CANNOT create proxy instances of this entity and will ALWAYS load the entity eagerly.

I don't understand this part of the documentation. What is exactly preferable and what kind of performance issues can one expect if non-leaf entities are used?

Let's say we've got these entites:

```
#[InheritanceType('SINGLE_TABLE')]
class AbstractEngine {}

class V6Engine extends AbstractEngine {}
class V8Engine extends AbstractEngine {}
class AudiV8Engine extends V8Engine {}
```

Can I have a `Automobile` entity with a relationship with `AbstractEngine`? Or should it be with a leaf entity, so only `V6Engine` or `AudiV8Engine` can be used?

**Comment by greg0ire** [comment, marked as answer]
> Can I have a Automobile entity with a relationship with AbstractEngine?

Yes, it can, but there will be a performance hit you would not have if you were using a leaf entity: the cost in performance is the cost of hydrating the engine, from what I understand. When you instantiate an `Automobile`, the ORM cannot create a proxy instance for the engine, so it will:
- perform an SQL request to get the engine's data
- hydrate an instance of `AbstractEngine` with that data

If you are instead targeting a leaf class, the query + hydration would only happen when doing `$automobile->engine()->name()`. In the first case, it happens when doing `new Automobile()`.

While there is a performance hit, it's probably pretty mild though, but only you can tell.

**Reply by beberlei** [reply]
The performance hit often comes from N+1 queries. Say you have a list of 20 automobiles, then it will trigger 20 queries for the `AbstractEngine`.
